### PR TITLE
Metric Error Codes Recording + Error Code Prefixing

### PR DESCRIPTION
**Pull request by jake-engelberg**
# Description

Error codes that were defined and used on a per-developer basis for debug logging has been temporarily moved to a singular `errorcodes` package to be defined and prefixed for cohesion. Eventually, they will be converted to structures similar to `predefined.go` for all errors and associated messages.

Additionally, error codes are now retrieved using a function which also records a count metric for tracking errors.

No renaming has been done to assure backwards compatability.

The metric can be filtered by its keys for app-id, error code, and error category.

The error metric recording is opt-in via the flag `metrics.recordErrorCodes`.

Metric fields/schema:
```
dapr_error_code_total
{
    app_id = "errorgen-app",
    category = "pubsub",
    dapr_io_enabled = "true",
    error_code = "ERR_PUBSUB_NOT_FOUND",
    instance = "10.244.0.35:9090",
    job = "kubernetes-service-endpoints",
    namespace = "crud-app",
    node = "dapr-master-control-plane",
    service = "errorgen-app-dapr"
}
```

Example of query `dapr_error_code_total` in Grafana:
Screenshot 2024-11-14 at 10 00 41 PM

Example of query `rate(dapr_error_code_total[2m])` in Grafana:
Screenshot 2024-11-14 at 10 00 12 PM

Another img:
Screenshot 2024-11-27 at 11 44 37 AM

## Issue reference

Please reference the issue this PR will close: https://github.com/dapr/proposals/pull/67

## Checklist

Please make sure you've completed the relevant tasks for this PR, out of the following list:

- [x] Code compiles correctly
- [x] Created/updated tests
- [x] Unit tests passing
- [x] End-to-end tests passing
- [x] Extended the documentation / Created issue in the <https://github.com/dapr/docs/> repo: https://github.com/dapr/docs/pull/4422
- [x] Specification has been updated / Created issue in the <https://github.com/dapr/docs/> repo: https://github.com/dapr/docs/pull/4422
- [x] Provided sample for the feature / Created issue in the <https://github.com/dapr/docs/> repo: https://github.com/dapr/docs/pull/4422

### CODE REVIEW

File: tests/e2e/healthcheck/healthcheck_test.go
Comment (filintod) [inline tests/e2e/healthcheck/healthcheck_test.go:269]: let's use the constant value from `errorcodes` package. the same in all tests

Comment (elena-kolevska) [reply tests/e2e/healthcheck/healthcheck_test.go]: I'm in the camp of always using hardcoded error codes in tests, so that it's very obvious if we break compatibility in a later version.

**Review by elena-kolevska — CHANGES_REQUESTED** [review]
I'll be doing a more detailed review tomorrow, but sending my comments about breaking changes early so you can work on it while we're looking at the rest.

File: pkg/api/http/directmessaging_test.go
Comment (jake-engelberg) [reply pkg/api/http/directmessaging_test.go:414]: I'm aware of this, and I don't exactly have much of a preference of whether or not to change the errors; It was described to me as breaking but logs are not in the contract of what we have to be super weary of; the old error codes were inconsistent in naming, like:
ERR_RAISE_EVENT_WORKFLOW
ERR_WORKFLOW_NAME_MISSING
...so I envision someday we'd want to rename these errors, unless you see errors like ERR_WORKFLOW_NAME_MISSING not as being prefixed but just a phrase that happens to start with the component.

In the case of the latter, as well as generally just wanting to prevent breaking changes, I can revert to the original error codes.

Comment (elena-kolevska) [reply pkg/api/http/directmessaging_test.go:414]: Aren't these the same codes that are returned in http responses too? I haven't looked deeper into the PR yet, this is just something that caught my eye at first glance.

Comment (jake-engelberg) [reply pkg/api/http/directmessaging_test.go:414]: Sometimes, yes; in that case, it would be more a potential breaking change if users rely on consistent responses, so it adds another reason to revert to original error codes which I will do so.

File: pkg/config/testdata/override_gen.yaml
Comment (JoshVanL) [inline pkg/config/testdata/override_gen.yaml:28]:
```suggestion
        recordErrorCodes: null
```

File: pkg/config/testdata/override_spec_gen.yaml
Comment (JoshVanL) [inline pkg/config/testdata/override_spec_gen.yaml:9]:
```suggestion
    recordErrorCodes: null
```

File: pkg/diagnostics/errorcode_monitoring.go
Comment (JoshVanL) [inline pkg/diagnostics/errorcode_monitoring.go:1]: Add copyright headers.

File: pkg/diagnostics/errorcode_monitoring.go
Comment (JoshVanL) [inline pkg/diagnostics/errorcode_monitoring.go:18]:
```suggestion
```

File: pkg/diagnostics/errorcode_monitoring.go
Comment (JoshVanL) [inline pkg/diagnostics/errorcode_monitoring.go:30]: AppID?

File: pkg/diagnostics/errorcode_monitoring_test.go
Comment (JoshVanL) [inline pkg/diagnostics/errorcode_monitoring_test.go:1]: Copyright headers.

File: pkg/diagnostics/errorcode_monitoring.go
Comment (JoshVanL) [inline pkg/diagnostics/errorcode_monitoring.go:42]: Remove, and move to `New`.

File: pkg/messages/errorcodes/errorcodes.go
Comment (JoshVanL) [inline pkg/messages/errorcodes/errorcodes.go:2]:
```suggestion
Copyright 2024 The Dapr Authors
```

**Review by JoshVanL — CHANGES_REQUESTED** [review]

File: pkg/diagnostics/service_monitoring.go
Comment (nelson-parente) [inline pkg/diagnostics/service_monitoring.go:26]: nit: follow the same pattern we have on other keys error_code or errorCode (we have both, choose one I guess)

File: pkg/config/configuration.go
Comment (nelson-parente) [inline pkg/config/configuration.go:258]: add yaml tag

File: pkg/api/http/metadata.go
Comment (nelson-parente) [inline pkg/api/http/metadata.go:25]: nit: remove this new line

File: pkg/diagnostics/errorcode_monitoring.go
Comment (jake-engelberg) [reply pkg/diagnostics/errorcode_monitoring.go:30]: app-id is supplied with the `Init()` function, following the other metrics' structures
https://github.com/jake-engelberg/dapr//blob/5c3b42bf03519008739c1fc2ec8030dba5f1fd00/pkg/diagnostics/errorcode_monitoring.go#L48

File: pkg/diagnostics/errorcode_monitoring.go
Comment (jake-engelberg) [reply pkg/diagnostics/errorcode_monitoring.go:42]: the metric is optional and so the Init() is a secondary function to call when the flag is checked

https://github.com/jake-engelberg/dapr//blob/5c3b42bf03519008739c1fc2ec8030dba5f1fd00/pkg/diagnostics/metrics.go#L53

**Review by elena-kolevska — CHANGES_REQUESTED** [review]
Thank you @jake-engelberg I'm happy we're moving even further in making Dapr errors easier to work with. 🚀 

I left a few comments, but I also wanted to point something that's missing: 
The current error metric recording relies on the error code we used to use for http mostly, but we don't have that for new code like the scheduler and the Jobs API where we use the Rich grpc errors exclusively and we use the `ErrorInfo` string to derive the http error code (you can see how that's done [here](https://github.com/dapr/kit/blob/main/errors/errors.go#L214)).
It's a simple change though, you can look through the [scheduler errors file](https://github.com/dapr/dapr/blob/master/pkg/api/errors/scheduler.go) and look at the `WithErrorInfo` calls. The first argument is the tag you need to record.

We'll also need to do this for the control plane services, but we can leave that for a separate PR.

**Comment by jake-engelberg** [comment]
> Thank you @jake-engelberg I'm happy we're moving even further in making Dapr errors easier to work with. 🚀
> 
> I left a few comments, but I also wanted to point out that we seem not to be recording scheduler errors; The current error metric recording relies on the error code we used to use for http mostly, but we don't have that for new code like the scheduler and the Jobs API where we use the Rich grpc errors exclusively and we use the `ErrorInfo` string to derive the http error code (you can see how that's done [here](https://github.com/dapr/kit/blob/main/errors/errors.go#L214)). It's a simple change though, you can look through the [scheduler errors file](https://github.com/dapr/dapr/blob/master/pkg/api/errors/scheduler.go) and look at the `WithErrorInfo` calls. The first argument is the tag you need to record.
> 
> We'll also need to do this for the control plane services, but we can leave that for a separate PR.

Thanks for this point brought up; I skipped over the Jobs API errors because of those composite codes, but I went back and made some specific functions to allow recording of those various codes even though they may not be defined in the new `errorcodes` package.

While I would have liked to either modify WithErrorInfo, or wrap it, to minimize repetition, I decided against it for these two reasons:
1. I wanted to keep dependencies within `dapr/dapr`, so I decided to not make changes to WithErrorInfo in `dapr/kit`

2. Within `dapr/dapr`, I can't make a wrapper to record and call `WithErrorInfo()` as well as `RecordAndGet()` as the calling struct is defined in `dapr/kit`, not local to anywhere in `dapr/dapr`, so I can't do:
```
func (b *kiterrors.ErrorBuilder) WithErrorInfoAndRecord(){
    //Performs metric recording as well as has original WithErrorInfo() function call
}
```

LMK if you see some other way to reduce repetition, for my solution meant adding a `RecordAndGetComp()` call with each `WithErrorInfo()`.

**Comment by elena-kolevska** [comment]
@jake-engelberg  Fix linter please, so the tests can run 🙏

**Comment by filintod** [comment]
could we change the suffix to be `_total` instead of `_count` to follow convention https://prometheus.io/docs/practices/naming/. We don't follow a single convention in dapr but probably to move to common ones as we go

File: pkg/api/http/actors.go
Comment (jake-engelberg) [reply pkg/api/http/actors.go]: Changes related to above: https://github.com/dapr/dapr/pull/8256/commits/7c6a69175d553c82f11b99f847ca5108000e3a9b

I've went through and pushed the metric recording up in the call stack as far as I could understand, and ended up with a few functions that take different structures depending on how the end of the chain's parameters were / how the error's structure ended up looking like:

1. `RecordAPIErrorCode()`
    * Used at the end of most HTTP responses, which are provided an APIError struct
2. `RecordErrorCode()`
    * Used at the end of various chains, which are provided the new ErrorCode struct, like some other HTTP responses, the HTTP error response helper `NewErrorResponse()`, some crypto gRPC errors, Pubsub error builder
3. `RecordCompAndGet()`
    * Used at the end of various chains, which are provided various composited errors not following the `"ERR_X_X_X" format, some general error builders, scheduler errors, state error builder


In my opinion, it might be somewhat less understandable than having the recording added to each usage of the error codes, but it did lower the lines changed drastically and seems like a workable solution for the time being.

File: pkg/diagnostics/errorcode_monitoring_test.go
Comment (jake-engelberg) [reply pkg/diagnostics/errorcode_monitoring_test.go:31]: There is an additional test for 2 more error codes at the same time, can you expand on this suggestion?
Screenshot 2024-11-13 at 12 36 32 PM

**Review by elena-kolevska — CHANGES_REQUESTED** [review]
We're getting there @jake-engelberg , thanks for being patient with us. It's so important we get this right :)
I have to say personally I don't love the record and get thing. A method with two verbs is usually a code smell and it violates single responsabillity principle, but I'm not going to hold on approval based on that (as soon as we clean everything up).

We also need to wait for [this PR](https://github.com/dapr/dapr/pull/8237) to get merged, because we're gonna have to adjust some code.

File: pkg/api/universal/subtlecrypto_subtlecrypto.go
Comment (jake-engelberg) [reply pkg/api/universal/subtlecrypto_subtlecrypto.go:161]: My understanding of the call hiearchy is that the functions in this file are handled by `UniversalHTTPHandler()`
Screenshot 2024-11-14 at 3 07 19 PM

Which then handles lots of APIErrors already but also the one's you commented about here.
Screenshot 2024-11-14 at 3 08 39 PM

File: pkg/api/errors/errors.go
Comment (jake-engelberg) [reply pkg/api/errors/errors.go:43]: I think it actually is because it ends up at `UniversalHTTPHandler()` like [this comment as well](https://github.com/dapr/dapr/pull/8256#discussion_r1842830509), I'll remove.

File: pkg/api/universal/actors.go
Comment (jake-engelberg) [reply pkg/api/universal/actors.go:58]: `respondWithError()`
Screenshot 2024-11-14 at 3 23 48 PM

File: pkg/api/grpc/crypto.go
Comment (jake-engelberg) [reply pkg/api/grpc/crypto.go]: You are correct, I made a wrapper to record any of the returned APIErrors for both, lmk if this is bad practice:
Screenshot 2024-11-14 at 3 25 10 PM

File: pkg/api/universal/actors.go
Comment (elena-kolevska) [reply pkg/api/universal/actors.go:58]: That's only on the http side though, no? If the universal API is accessed through the gRPC side, it won't record it.
Please feel free to correct me if I'm missing something, you've been looking into this way more than me.

Comment (jake-engelberg) [reply pkg/api/universal/actors.go:58]: This is correct, I ended up adding to the gRPC metric recording which is a middleware that will intercept the errors to do the same thing as the universal HTTP recorder.

**Comment by elena-kolevska** [comment]
Hey @jake-engelberg  the PR I referred to before is merged, please update your branch 🙏 

i saw the changes you made and i think we're going in the right direction, by centralising the recording as much as possible.

**Comment by jake-engelberg** [comment]
https://github.com/dapr/kit/pull/109

This PR makes some changes to dapr/kit that are necessary for the latest commit I made.

File: pkg/api/http/responses.go
Comment (elena-kolevska) [inline pkg/api/http/responses.go:127]: Seems like we're not recording this error

File: pkg/components/pluggable/errors.go
Comment (elena-kolevska) [inline pkg/components/pluggable/errors.go:116]: We need to look a bit deeper into this, if I remember correctly how pluggable components work, we won't be able to extract the original category, but we could maybe create a new generic one, for example, `pluggable-state`.  @JoshVanL what do you think?
For context (so you don't have to look through the whole PR), we added a `category` field needed for recording metrics to kit and APIErrors, but we never return it as part of the final error the client sees. Does that mean a pluggable component won't have access to it?

Comment (elena-kolevska) [reply pkg/components/pluggable/errors.go:116]: I'm also ok marking this as category `pluggable-component` for now, and split it in the future if necessary.

File: pkg/diagnostics/grpc_monitoring.go
Comment (elena-kolevska) [inline pkg/diagnostics/grpc_monitoring.go:230]: Let's check if err is nil before calling the record function

File: tests/integration/framework/process/daprd/daprd.go
Comment (elena-kolevska) [inline tests/integration/framework/process/daprd/daprd.go:353]:
```suggestion

```

**Review by elena-kolevska — CHANGES_REQUESTED** [review]
Looks like we're almost done! Just a few small nits.

File: pkg/diagnostics/grpc_monitoring.go
Comment (filintod) [reply pkg/diagnostics/grpc_monitoring.go]: wdyt on adding a nil check inside `RecordErrorCode`?

like:
```
func RecordErrorCode(err error) bool {
  if err == nil { return false}
  ...
}
```

Comment (filintod) [reply pkg/diagnostics/grpc_monitoring.go]: but no need to change more in any case

Comment (elena-kolevska) [reply pkg/diagnostics/grpc_monitoring.go]: Yep, even better!

**Comment by elena-kolevska** [comment]
Looking great! Let's wait for the tests the pass and I'll approve. Good job Jake!

**Review by elena-kolevska — DISMISSED** [review]

**Review by elena-kolevska — APPROVED** [review]

**Review by artursouza — APPROVED** [review]

**Comment by marcduiker** [comment]
@holopin-bot @jake-engelberg Thanks Jake!

**Pull request merged**
